File: android/src/main/java/vn/reactnativebspatch/OtaBundleResolver.kt
package vn.reactnativebspatch

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * OtaBundleResolver — encapsulates OTA bundle path resolution with
 * crash detection and automatic rollback.
 *
 * IMPORTANT: This runs BEFORE any JS code. If the OTA bundle is corrupted,
 * Hermes crashes before React Native boots. This native resolver handles
 * crash detection so the app can recover without JS ever running.
 *
 * Lifecycle:
 *   1. Read metadata.json
 *   2. If lastLaunchSuccessful == false → rollback to previous version
 *   3. Mark current launch as pending (lastLaunchSuccessful = false)
 *   4. Return the bundle path (or null for APK default)
 *   5. JS calls markAsHealthy() → sets lastLaunchSuccessful = true
 *   6. If crash before markAsHealthy → next boot sees false → rollback
 *
 * Usage in MainApplication.kt:
 * ```
 * jsBundleFilePath = OtaBundleResolver.resolve(applicationContext)
 * ```
 */
object OtaBundleResolver {

    private const val TAG = "OtaBundleResolver"
    private const val OTA_DIR_NAME = "ota"
    private const val METADATA_FILE = "metadata.json"
    private const val DEFAULT_BASE_VERSION = 1
    private const val MAX_ROLLBACK = 3

    /**
     * Resolves the JS bundle path from OTA metadata.
     * Handles crash detection and rollback entirely on the native side.
     *
     * @param context Android Application context
     * @param baseVersion The OTA version baked into the APK (default: 1)
     * @return Absolute path to the OTA bundle, or null to use the APK default.
     */
    @JvmStatic
    fun resolve(context: Context, baseVersion: Int = DEFAULT_BASE_VERSION): String? {
        return try {
            val otaDir = File(context.filesDir, OTA_DIR_NAME)
            val metadataFile = File(otaDir, METADATA_FILE)

            if (!metadataFile.exists()) {
                Log.d(TAG, "No OTA metadata → APK bundle")
                return null
            }

            var metadata = JSONObject(metadataFile.readText())
            val currentVersion = metadata.optInt("currentVersion", baseVersion)
            val lastLaunchOk = metadata.optBoolean("lastLaunchSuccessful", true)
            val rollbackCount = metadata.optInt("rollbackCount", 0)

            // ── Step 1: Crash detection & rollback ──────────────────────────
            if (!lastLaunchOk && currentVersion > baseVersion) {
                Log.w(TAG, "Previous launch failed! Rolling back from V$currentVersion...")
                metadata = performRollback(metadata, baseVersion, rollbackCount, otaDir)
                writeMetadata(metadataFile, metadata)

                val newVersion = metadata.optInt("currentVersion", baseVersion)
                if (newVersion <= baseVersion) {
                    Log.i(TAG, "Rolled back to factory APK bundle")
                    return null
                }

                // Check if the rollback target bundle exists
                val rollbackVDir = File(otaDir, "v${newVersion}")
                val rollbackBundle = File(rollbackVDir, "index.android.bundle")
                if (!rollbackBundle.exists()) {
                    Log.w(TAG, "Rollback target index.android.bundle missing in V$newVersion → factory")
                    resetToFactory(metadataFile, baseVersion)
                    return null
                }

                // Mark this rollback attempt as pending
                metadata.put("lastLaunchSuccessful", false)
                writeMetadata(metadataFile, metadata)

                Log.i(TAG, "Trying rollback bundle: ${rollbackBundle.absolutePath}")
                return rollbackBundle.absolutePath
            }

            // ── Step 2: Normal boot — no OTA bundle needed ──────────────────
            if (currentVersion <= baseVersion) {
                Log.d(TAG, "OTA v$currentVersion ≤ base v$baseVersion → APK bundle")
                return null
            }

            // ── Step 3: OTA bundle — mark as pending verification ───────────
            val vDir = File(otaDir, "v${currentVersion}")
            val bundleFile = File(vDir, "index.android.bundle")
            if (!bundleFile.exists()) {
                Log.w(TAG, "index.android.bundle missing in V$currentVersion → APK bundle")
                resetToFactory(metadataFile, baseVersion)
                return null
            }

            // Mark launch as pending — if Hermes crashes loading this bundle,
            // the next boot will see false and trigger rollback
            metadata.put("lastLaunchSuccessful", false)
            writeMetadata(metadataFile, metadata)

            Log.i(TAG, "Loading OTA bundle: ${bundleFile.absolutePath}")
            bundleFile.absolutePath
        } catch (e: Exception) {
            Log.e(TAG, "Error resolving OTA bundle: ${e.message}")
            null
        }
    }

    /**
     * Performs rollback by popping from previousVersions.
     * If maxRollback reached, resets to factory.
     */
    private fun performRollback(
        metadata: JSONObject,
        baseVersion: Int,
        currentRollbackCount: Int,
        otaDir: File
    ): JSONObject {
        val newCount = currentRollbackCount + 1

        // Max rollback reached → factory reset
        if (newCount >= MAX_ROLLBACK) {
            Log.w(TAG, "Max rollback ($MAX_ROLLBACK) reached → factory reset")
            return JSONObject().apply {
                put("currentVersion", baseVersion)
                put("previousVersions", JSONArray())
                put("rollbackCount", 0)
                put("lastLaunchSuccessful", true)
            }
        }

        val prevVersions = metadata.optJSONArray("previousVersions") ?: JSONArray()

        if (prevVersions.length() == 0) {
            Log.w(TAG, "No previous versions to rollback to → factory reset")
            return JSONObject().apply {
                put("currentVersion", baseVersion)
                put("previousVersions", JSONArray())
                put("rollbackCount", newCount)
                put("lastLaunchSuccessful", true)
            }
        }

        // Pop the first previous version
        val rollbackTo = prevVersions.optInt(0, baseVersion)
        val remaining = JSONArray()
        for (i in 1 until prevVersions.length()) {
            remaining.put(prevVersions.optInt(i))
        }

        Log.i(TAG, "Rolling back: V${metadata.optInt("currentVersion")} → V$rollbackTo (attempt $newCount)")

        // Check if this rollback target bundle exists
        val vDir = File(otaDir, "v${rollbackTo}")
        val targetBundle = File(vDir, "index.android.bundle")
        if (!targetBundle.exists() && rollbackTo > baseVersion) {
            Log.w(TAG, "index.android.bundle not found in V$rollbackTo, trying deeper rollback...")
            val intermediate = JSONObject().apply {
                put("currentVersion", rollbackTo)
                put("previousVersions", remaining)
                put("rollbackCount", newCount)
                put("lastLaunchSuccessful", false)
            }
            return performRollback(intermediate, baseVersion, newCount, otaDir)
        }

        return JSONObject().apply {
            put("currentVersion", rollbackTo)
            put("previousVersions", remaining)
            put("rollbackCount", newCount)
            put("lastLaunchSuccessful", true) // Will be set to false again in resolve()
        }
    }

    /**
     * Resets metadata to factory defaults.
     */
    private fun resetToFactory(metadataFile: File, baseVersion: Int) {
        val factory = JSONObject().apply {
            put("currentVersion", baseVersion)
            put("previousVersions", JSONArray())
            put("rollbackCount", 0)
            put("lastLaunchSuccessful", true)
        }
        writeMetadata(metadataFile, factory)
    }

    @JvmStatic
    fun markAsHealthy(context: Context) {
        try {
            val otaDir = File(context.filesDir, OTA_DIR_NAME)
            val metadataFile = File(otaDir, METADATA_FILE)
            if (!metadataFile.exists()) return

            val metadata = JSONObject(metadataFile.readText())
            metadata.put("lastLaunchSuccessful", true)
            metadata.put("rollbackCount", 0)

            writeMetadata(metadataFile, metadata)
            Log.i(TAG, "App marked as healthy")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to mark as healthy: ${e.message}")
        }
    }

    /**
     * Writes metadata JSON to disk.
     */
    private fun writeMetadata(file: File, metadata: JSONObject) {
        try {
            file.parentFile?.mkdirs()
            file.writeText(metadata.toString(2))
        } catch (e: Exception) {
            Log.e(TAG, "Failed to write metadata: ${e.message}")
        }
    }
}

File: ios/OtaBundleResolver.swift
import Foundation

/**
 * OtaBundleResolver — encapsulates OTA bundle path resolution for iOS with
 * crash detection and automatic rollback.
 *
 * IMPORTANT: This runs BEFORE any JS code. If the OTA bundle is corrupted,
 * Hermes crashes before React Native boots. This native resolver handles
 * crash detection so the app can recover without JS ever running.
 *
 * Lifecycle:
 *   1. Read metadata.json
 *   2. If lastLaunchSuccessful == false → rollback to previous version
 *   3. Mark current launch as pending (lastLaunchSuccessful = false)
 *   4. Return the bundle path (or nil for IPA default)
 *   5. JS calls markAsHealthy() → sets lastLaunchSuccessful = true
 *   6. If crash before markAsHealthy → next boot sees false → rollback
 *
 * Usage in AppDelegate.swift:
 * ```swift
 * override func bundleURL() -> URL? {
 *     if let otaBundle = OtaBundleResolver.resolve() {
 *         return otaBundle
 *     }
 *     // Default RCTBundleURLProvider ...
 * }
 * ```
 */
@objcMembers
public class OtaBundleResolver: NSObject {

    private static let otaDirName = "ota"
    private static let metadataFile = "metadata.json"
    public static let defaultBaseVersion = 1
    private static let maxRollback = 3
    private static let maxRollbackIterations = 10 // Safety limit to prevent infinite loops

    /// Resolves the JS bundle path from OTA metadata.
    /// Handles crash detection and rollback entirely on the native side.
    ///
    /// - Parameter baseVersion: The OTA version baked into the IPA (default: 1)
    /// - Returns: URL to the OTA bundle, or nil to use the IPA default.
    @objc public static func resolve(baseVersion: Int = defaultBaseVersion) -> URL? {
        guard let documentsDir = FileManager.default.urls(
            for: .documentDirectory,
            in: .userDomainMask
        ).first else {
            return nil
        }

        let otaDir = documentsDir.appendingPathComponent(otaDirName)
        let metadataURL = otaDir.appendingPathComponent(metadataFile)

        guard FileManager.default.fileExists(atPath: metadataURL.path) else {
            NSLog("[OtaBundleResolver] No OTA metadata → IPA bundle")
            return nil
        }

        do {
            let data = try Data(contentsOf: metadataURL)
            guard var metadata = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return nil
            }

            let currentVersion = metadata["currentVersion"] as? Int ?? baseVersion
            let lastLaunchOk = metadata["lastLaunchSuccessful"] as? Bool ?? true
            let rollbackCount = metadata["rollbackCount"] as? Int ?? 0

            // ── Step 1: Crash detection & rollback ──────────────────────────
            if !lastLaunchOk && currentVersion > baseVersion {
                NSLog("[OtaBundleResolver] Previous launch failed! Rolling back from V\(currentVersion)...")
                
                metadata = performRollback(metadata: metadata, baseVersion: baseVersion, currentRollbackCount: rollbackCount, otaDir: otaDir)
                writeMetadata(url: metadataURL, metadata: metadata)

                let newVersion = metadata["currentVersion"] as? Int ?? baseVersion
                if newVersion <= baseVersion {
                    NSLog("[OtaBundleResolver] Rolled back to factory IPA bundle")
                    return nil
                }

                let rollbackVDir = otaDir.appendingPathComponent("v\(newVersion)")
                let rollbackBundle = rollbackVDir.appendingPathComponent("main.jsbundle")
                if !FileManager.default.fileExists(atPath: rollbackBundle.path) {
                    NSLog("[OtaBundleResolver] Rollback target main.jsbundle missing in V\(newVersion) → factory")
                    resetToFactory(url: metadataURL, baseVersion: baseVersion)
                    return nil
                }

                // Mark this rollback attempt as pending
                metadata["lastLaunchSuccessful"] = false
                writeMetadata(url: metadataURL, metadata: metadata)

                NSLog("[OtaBundleResolver] Trying rollback bundle: \(rollbackBundle.path)")
                return rollbackBundle
            }

            // ── Step 2: Normal boot — no OTA bundle needed ──────────────────
            if currentVersion <= baseVersion {
                NSLog("[OtaBundleResolver] OTA v\(currentVersion) ≤ base v\(baseVersion) → IPA bundle")
                return nil
            }

            // ── Step 3: OTA bundle — mark as pending verification ───────────
            let vDir = otaDir.appendingPathComponent("v\(currentVersion)")
            let bundleURL = vDir.appendingPathComponent("main.jsbundle")
            if !FileManager.default.fileExists(atPath: bundleURL.path) {
                NSLog("[OtaBundleResolver] main.jsbundle missing in V\(currentVersion) → IPA bundle")
                resetToFactory(url: metadataURL, baseVersion: baseVersion)
                return nil
            }

            // Mark launch as pending
            metadata["lastLaunchSuccessful"] = false
            writeMetadata(url: metadataURL, metadata: metadata)

            NSLog("[OtaBundleResolver] Loading OTA bundle: \(bundleURL.path)")
            return bundleURL

        } catch {
            NSLog("[OtaBundleResolver] Error: \(error.localizedDescription)")
            return nil
        }
    }

    private static func performRollback(metadata: [String: Any], baseVersion: Int, currentRollbackCount: Int, otaDir: URL) -> [String: Any] {
        var rollbackCount = currentRollbackCount + 1
        var prevVersions = metadata["previousVersions"] as? [Int] ?? []
        var current = metadata
        var iteration = 0

        while iteration < maxRollbackIterations {
            iteration += 1

            if rollbackCount >= maxRollback {
                NSLog("[OtaBundleResolver] Max rollback (\(maxRollback)) reached → factory reset")
                return [
                    "currentVersion": baseVersion,
                    "previousVersions": [Int](),
                    "rollbackCount": 0,
                    "lastLaunchSuccessful": true
                ]
            }

            guard !prevVersions.isEmpty else {
                NSLog("[OtaBundleResolver] No previous versions to rollback to → factory reset")
                return [
                    "currentVersion": baseVersion,
                    "previousVersions": [Int](),
                    "rollbackCount": rollbackCount,
                    "lastLaunchSuccessful": true
                ]
            }

            let rollbackTo = prevVersions[0]
            prevVersions = Array(prevVersions.dropFirst())

            NSLog("[OtaBundleResolver] Rolling back: V\(current["currentVersion"] as? Int ?? baseVersion) → V\(rollbackTo) (attempt \(rollbackCount))")

            let vDir = otaDir.appendingPathComponent("v\(rollbackTo)")
            let targetBundle = vDir.appendingPathComponent("main.jsbundle")
            if !FileManager.default.fileExists(atPath: targetBundle.path) && rollbackTo > baseVersion {
                NSLog("[OtaBundleResolver] main.jsbundle not found in V\(rollbackTo), trying deeper rollback...")
                current = [
                    "currentVersion": rollbackTo,
                    "previousVersions": prevVersions,
                    "rollbackCount": rollbackCount,
                    "lastLaunchSuccessful": false
                ]
                rollbackCount += 1
                continue
            }

            return [
                "currentVersion": rollbackTo,
                "previousVersions": prevVersions,
                "rollbackCount": rollbackCount,
                "lastLaunchSuccessful": true
            ]
        }

        // Safety: if we somehow exit the loop without returning, do factory reset
        NSLog("[OtaBundleResolver] Max rollback iterations (\(maxRollbackIterations)) exceeded → factory reset")
        return [
            "currentVersion": baseVersion,
            "previousVersions": [Int](),
            "rollbackCount": 0,
            "lastLaunchSuccessful": true
        ]
    }

    private static func resetToFactory(url: URL, baseVersion: Int) {
        let factory: [String: Any] = [
            "currentVersion": baseVersion,
            "previousVersions": [Int](),
            "rollbackCount": 0,
            "lastLaunchSuccessful": true
        ]
        writeMetadata(url: url, metadata: factory)
    }

    @objc public static func markAsHealthy() {
        guard let documentsDir = FileManager.default.urls(
            for: .documentDirectory, 
            in: .userDomainMask
        ).first else {
            return
        }

        let otaDir = documentsDir.appendingPathComponent(otaDirName)
        let metadataURL = otaDir.appendingPathComponent(metadataFile)

        guard FileManager.default.fileExists(atPath: metadataURL.path) else {
            return
        }

        do {
            let data = try Data(contentsOf: metadataURL)
            guard var metadata = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return
            }

            metadata["lastLaunchSuccessful"] = true
            metadata["rollbackCount"] = 0
            
            writeMetadata(url: metadataURL, metadata: metadata)
            NSLog("[OtaBundleResolver] App marked as healthy")
        } catch {
            NSLog("[OtaBundleResolver] Failed to mark as healthy: \(error.localizedDescription)")
        }
    }

    private static func writeMetadata(url: URL, metadata: [String: Any]) {
        do {
            let data = try JSONSerialization.data(withJSONObject: metadata, options: .prettyPrinted)
            try data.write(to: url)
        } catch {
            NSLog("[OtaBundleResolver] Failed to write metadata: \(error.localizedDescription)")
        }
    }
}
